src/app/teacher/apply-tags-button/apply-tags-button.component.ts

```diff
@@ -19,21 +21,37 @@ export class ApplyTagsButtonComponent implements OnInit {
   constructor(private dialog: MatDialog, private projectTagService: ProjectTagService) {}
 
   ngOnInit(): void {
-    this.retrieveUserTags();
     this.subscribeToTagUpdated();
     this.subscribeToNewTag();
     this.subscribeToTagDeleted();
   }
 
+  ngOnChanges(): void {
+    this.retrieveUserTags();
+  }
+
   private retrieveUserTags(): void {
     this.projectTagService.retrieveUserTags().subscribe((tags: Tag[]) => {
       for (const tag of tags) {
-        tag.checked = this.doesAnyProjectHaveTag(tag);
+        this.updateTagCheckedValue(tag);
       }
       this.tags = tags;
     });
   }
 
+  private updateTagCheckedValue(tag: Tag): void {
+    const numProjectsWithTag = this.getNumProjectsWithTag(tag);
+    tag.checked = numProjectsWithTag === this.selectedProjects.length;
+    tag.indeterminateChecked =
+      numProjectsWithTag > 0 && numProjectsWithTag < this.selectedProjects.length;
+  }
+
+  private getNumProjectsWithTag(tag: Tag): number {
+    return this.selectedProjects.filter((project) =>
+      project.tags.some((projectTag) => projectTag.id === tag.id)
+    ).length;
+  }
```

Comment:
Do we need to define a function here, or can we just call ```this.selectedProjects.filter(...)``` inline in updateTagCheckedValue()?

Reply:
Ok, I've made this change.